drivers/flash/flash_util.c

```diff
@@ -80,3 +87,199 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
 	return -ENOSYS;
 #endif
 }
+
+static bool off_add_overflow(off_t offset, off_t size, off_t *result)
+{
+	if (sizeof(off_t) == sizeof(uint32_t)) {
+		uint32_t end;
+
+		/* account for signedness of off_t due to lack of s32_add_overflow() */
+		if (u32_add_overflow((uint32_t)offset, (uint32_t)size, &end) ||
+		    (end > (UINT32_MAX >> 1))) {
+			return true;
+		}
+	} else if (sizeof(off_t) == sizeof(uint64_t)) {
+		uint64_t end;
+
+		/* account for signedness of off_t due to lack of s64_add_overflow() */
+		if (u64_add_overflow((uint64_t)offset, (uint64_t)size, &end) ||
+		    (end > (UINT64_MAX >> 1))) {
+			return true;
+		}
+	}
```

Comment:
This is overly complicated.
Since `off_t` is signed, then 
```
#define OFF_T_MIN ((off_t)1 << (sizeof(off_t) * 8  -1))
#define OFF_T_MIN (~(OFF_T_MIN))
/* At this point both offset and  size should be >= 0, as they should be verified by caller */
if (OFF_T_MAX - offset < size) {
    return true;
}

*result = offset + size;
return false;
```

then there is the question whether the function is even needed as this can just be inlined.
This is the same approach as taken in Integer Security check from "Secure Coding in C and C++" or chapter INT32-C from "The CERT C Secure Coding Standard".
There should be check for offset >= 0 and size >=0 but this is assumed.

Reply:
I haven't yet addressed this, but will take a look again tomorrow most likely.